src/waiter.c

```diff
@@ -0,0 +1,263 @@
+
+typedef struct OP_CONTEXT_TAG
+{
+    int32_t initial_state;
+    THANDLE(WAITER) waiter;
+} OP_CONTEXT;
+
+void waiter_dispose(WAITER* waiter)
+{
+    int32_t state = interlocked_add(&waiter->state, 0);
+    if(state & WAITER_REGISTERED)
+    {
+        waiter->notification_callback(waiter->context, NULL, WAITER_CALLBACK_RESULT_ABANDONED);
+    }
+    else if (state & WAITER_NOTIFIED)
+    {
+        waiter->notify_complete_callback(waiter->context, WAITER_CALLBACK_RESULT_ABANDONED);
+        THANDLE_ASSIGN(RC_PTR)(&waiter->data, NULL);
+    }
+}
+
+THANDLE(WAITER) waiter_create(void)
+{
+    THANDLE(WAITER) result = THANDLE_MALLOC(WAITER)(waiter_dispose);
+    if (result == NULL)
+    {
+        LogError("Failure in THANDLE_MALLOC(WAITER)(waiter_dispose=%p)", waiter_dispose);
+    }
+    else
+    {
+        WAITER* waiter = THANDLE_GET_T(WAITER)(result);
+        waiter->notification_callback = NULL;
+        waiter->notify_complete_callback = NULL;
+        waiter->context = NULL;
+        THANDLE_INITIALIZE(RC_PTR)(&waiter->data, NULL);
+        (void)interlocked_exchange(&waiter->state, 0);
+    }
+    return result;
+}
+
+static void cancel_register_notification(void* context)
+{
+    if (context == NULL)
+    {
+        LogError("Invalid argument: void* context=%p", context);
+    }
+    else
+    {
+        OP_CONTEXT* op_context = context;
+        WAITER* waiter = THANDLE_GET_T(WAITER)(op_context->waiter);
+        if (interlocked_compare_exchange(&waiter->state, op_context->initial_state - WAITER_REGISTERED + WAITER_EPOCH_INCREMENT, op_context->initial_state) != op_context->initial_state)
+        {
+            LogError("Operation OP_CONTEXT* op_context = %p has been already cancelled or completed", op_context);
+        }
+        else
+        {
+            waiter->notification_callback(waiter->context, NULL, WAITER_CALLBACK_RESULT_CANCELLED);
+        }
+    }
+}
+
+static void dispose_async_op(void* context)
+{
+    if (context == NULL)
+    {
+        LogError("Invalid argument: void* context=%p", context);
+    }
+    else
+    {
+        OP_CONTEXT* op_context = context;
+        THANDLE_ASSIGN(WAITER)(&op_context->waiter, NULL);
+    }
+}
+
+WAITER_RESULT waiter_register_notification(THANDLE(WAITER) waiter, NOTIFICATION_CALLBACK notification_callback, void* context, THANDLE(ASYNC_OP)* out_op)
+{
+    WAITER_RESULT result;
+    if (
+        waiter == NULL ||
+        notification_callback == NULL ||
+        out_op == NULL
+        )
+    {
+        LogError("Invalid arguments: WAITER_HANDLE waiter = %p, NOTIFICATION_CALLBACK notification_callback = %p, void* context = %p, THANDLE(ASYNC_OP)* out_op = %p",
+            waiter, notification_callback, context, out_op);
+        result = WAITER_RESULT_ERROR;
+    }
+    else
+    {
+        WAITER* waiter_ptr = THANDLE_GET_T(WAITER)(waiter);
+        int32_t state = interlocked_or(&waiter_ptr->state, WAITER_REGISTERED);
+        if (state & WAITER_REGISTERED)
```

Comment:
so, in general when talking about interlocked and "state" *all* decision and inspection of the "state" should ensure that the state doesn't change in between 2 such reads.

Now in this case between lines 128 and 129 , nothing is prohibiting `waiter_notify` to be called. And I don't know what happens in that case, but I suppose somethign along inconsistency.

Also title of code review says "header and md" and those have been reviewed so... Iw ill stop here looking in the code.

Also for the next review (the one with code/tests/reals/UT/int) don't forget to add "chaos".

Reply:
And remove code in this iterations.

Reply:
I am changing my state to "waiting" because I think my comments might require some more work.